File: lib/permissions.js
const { BOT_OWNER_IDS } = require("./config");
const { gc } = require("../state/guildSettings");

function isOwner(idOrMember) {
  const id = typeof idOrMember === "string" ? idOrMember : idOrMember?.id;
  return BOT_OWNER_IDS.has(id);
}

function isMod(member) {
  if (!member) return false;
  if (isOwner(member)) return true;
  if (member.id === member.guild.ownerId) return true;
  const modRoleId = gc(member.guild).modRoleId;
  if (!modRoleId) return false;
  return member.roles.cache.has(modRoleId);
}

function isWhitelisted(member) {
  if (!member) return false;
  if (isOwner(member)) return true;                       // hardcoded owner is always immune
  if (member.id === member.guild.ownerId) return true;
  const g = gc(member.guild);
  if (g.nukeWhitelistUserIds.includes(member.id)) return true;
  return member.roles.cache.some(r => g.nukeWhitelistRoleIds.includes(r.id));
}

// Best-effort DM to a member before punitive action.
async function tryDM(user, text) {
  try { await user.send(text); } catch (_) {}
}

// Guard: can `actor` moderate `target`? Protects owner/whitelist and respects hierarchy.
function canActOn(actor, target) {
  if (!target) return { ok: false, why: "I can't find that user in this server." };
  if (isOwner(target)) return { ok: false, why: "That's the bot owner, so they're off-limits." };
  if (target.id === target.guild.ownerId) return { ok: false, why: "That's the server owner - can't touch them." };
  if (isWhitelisted(target)) return { ok: false, why: "That user's whitelisted, so they're protected." };
  if (target.id === actor.id) return { ok: false, why: "You can't do that to yourself." };
  const me = target.guild.members.me;
  if (me && target.roles.highest.position >= me.roles.highest.position)
    return { ok: false, why: "Their top role sits above mine, so I can't. Bump my role higher and try again." };
  const actorPrivileged = isOwner(actor) || actor.id === actor.guild.ownerId;
  if (!actorPrivileged && target.roles.highest.position >= actor.roles.highest.position)
    return { ok: false, why: "Their role is the same as or higher than yours, so this one's out of your reach." };
  return { ok: true };
}

module.exports = { isOwner, isMod, isWhitelisted, tryDM, canActOn };

File: rust/src/common/permissions.rs
//! Owner / mod / whitelist checks and the moderation hierarchy guard.

use serenity::builder::CreateMessage;
use serenity::http::Http;
use serenity::model::guild::Member;
use serenity::model::id::UserId;

use super::config::BOT_OWNER_IDS;
use super::guildinfo::GuildInfo;
use crate::state::guild_settings::gc;

pub fn is_owner(user_id: UserId) -> bool {
    BOT_OWNER_IDS.contains(&user_id.to_string())
}

pub fn is_mod(member: &Member, guild_owner_id: UserId) -> bool {
    if is_owner(member.user.id) {
        return true;
    }
    if member.user.id == guild_owner_id {
        return true;
    }
    let mod_role_id = gc(&member.guild_id.to_string()).mod_role_id;
    if mod_role_id.is_empty() {
        return false;
    }
    member.roles.iter().any(|r| r.to_string() == mod_role_id)
}

pub fn is_whitelisted(member: &Member, guild_owner_id: UserId) -> bool {
    if is_owner(member.user.id) {
        return true; // hardcoded owner is always immune
    }
    if member.user.id == guild_owner_id {
        return true;
    }
    let g = gc(&member.guild_id.to_string());
    if g.nuke_whitelist_user_ids.contains(&member.user.id.to_string()) {
        return true;
    }
    member.roles.iter().any(|r| g.nuke_whitelist_role_ids.contains(&r.to_string()))
}

/// Best-effort DM to a member before punitive action.
pub async fn try_dm(http: &Http, user_id: UserId, text: &str) {
    if let Ok(user) = user_id.to_user(http).await {
        let _ = user.direct_message(http, CreateMessage::new().content(text)).await;
    }
}

/// Guard: can `actor` moderate `target`? Protects owner/whitelist and respects
/// hierarchy. `Ok(())` means go ahead; `Err(why)` is the user-facing reason.
pub fn can_act_on(info: &GuildInfo, actor: &Member, target: &Member) -> Result<(), String> {
    if is_owner(target.user.id) {
        return Err("That's the bot owner, so they're off-limits.".into());
    }
    if target.user.id == info.owner_id {
        return Err("That's the server owner - can't touch them.".into());
    }
    if is_whitelisted(target, info.owner_id) {
        return Err("That user's whitelisted, so they're protected.".into());
    }
    if target.user.id == actor.user.id {
        return Err("You can't do that to yourself.".into());
    }
    let target_pos = info.member_highest(target);
    if info.bot_highest > 0 && target_pos >= info.bot_highest {
        return Err("Their top role sits above mine, so I can't. Bump my role higher and try again.".into());
    }
    let actor_privileged = is_owner(actor.user.id) || actor.user.id == info.owner_id;
    if !actor_privileged && target_pos >= info.member_highest(actor) {
        return Err("Their role is the same as or higher than yours, so this one's out of your reach.".into());
    }
    Ok(())
}
